src/cephadm/cephadm

```diff
@@ -107,6 +107,201 @@ class Ceph(object):
+            if ret:
+                logger.info('Enabling firewalld service %s in current zone...' % svc)
+                out, err, ret = call([self.cmd, '--permanent', '--add-service', svc])
+                if ret:
+                    raise RuntimeError(
+                        'unable to add service %s to current zone: %s' % (svc, err))
+            else:
+                logger.debug('firewalld service %s is enabled in current zone' % svc)
+
+    def open_ports(self, fw_ports):
+         # type (List[int]) -> None
+        if self.disabled:
+            return
+        if not self.available:
+            logger.debug('Not possible to open ports <%s>. firewalld.service is not available' % fw_ports)
+            return
+
+        for port in fw_ports:
+            tcp_port = str(port) + '/tcp'
+            out, err, ret = call([self.cmd, '--permanent', '--query-port', tcp_port], verbose_on_failure=False)
+            if ret:
+                logger.info('Enabling firewalld port %s in current zone...' % tcp_port)
+                out, err, ret = call([self.cmd, '--permanent', '--add-port', tcp_port])
+                if ret:
+                    raise RuntimeError('unable to add port %s to current zone: %s' %
+                                    (tcp_port, err))
+            else:
+                logger.debug('firewalld port %s is enabled in current zone' % tcp_port)
+
+    def apply_rules(self):
+         # type () -> None
+        if self.disabled or not self.available:
+            return
+
+        call_throws([self.cmd, '--reload'])
+
+class Cli(object):
+    def __init__(self, admin_key, uid, gid, config, log_dir, timeout=None):
+        # type: (str,int, int, str,str, Optional[int]) -> None
+        self.admin_key = admin_key
+        self.uid = uid
+        self.gid = gid
+        self.config = config
+        self.log_dir = log_dir
+        self.timeout = timeout
+
+    def ceph(self, cmd, extra_mounts={}, timeout=DEFAULT_TIMEOUT):
+        # type: (List[str], Dict[str, str], Optional[int]) -> str
+
+        tmp_admin_keyring = write_tmp('[client.admin]\n'
+                            '\tkey = ' + self.admin_key + '\n',
+                             self.uid, self.gid)
+        tmp_config = write_tmp(self.config, self.uid, self.gid)
+
+        mounts = { self.log_dir: '/var/log/ceph:z',
+                   tmp_admin_keyring.name: '/etc/ceph/ceph.client.admin.keyring:z',
+                   tmp_config.name: '/etc/ceph/ceph.conf:z',
+                 }
+        mounts.update(extra_mounts)
+        timeout = timeout or self.timeout
+
+        return CephContainer(
+            image=args.image,
+            entrypoint='/usr/bin/ceph',
+            args=cmd,
+            volume_mounts=mounts,
+        ).run(timeout=timeout)
+
+##################################
+
+class DashboardService(object):
+    """Defines the dashboard service when bootstrap a cluster"""
+    def __init__(self, args = None):
```

Comment:
what do you think of making `args` mandatory?

Reply:
It Is ok